File: algorithm/sort/Leetcode-147-InsertionSortList/Leetcode-147-InsertionSortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cur = ListNode()
        cur_node = cur 
        while head :
            while cur_node.next and cur_node.next.val < head.val :
                cur_node = cur_node.next
            temp = head.next
            temp2 = cur_node.next
            cur_node.next = head
            head.next = temp2
            head = temp

            cur_node = cur
        return cur.next

File: 5월 4주차/leetcode-147-Insertion Sort List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode beforeSortedHead(0);
        ListNode* beforeSortedHeadptr = &beforeSortedHead;

        while (head) {
            ListNode* beforecur = &beforeSortedHead;
            ListNode* cur = beforecur->next;
            while (true) {
                if (cur == nullptr || cur->val >= head->val) break;
                beforecur = beforecur->next;
                cur = cur->next;
            }
            ListNode* term = head;
            head = head->next;
            beforecur->next = term;
            term->next = cur;
        }
        return beforeSortedHead.next;
    }
};